src/core/net/dns_headers.cpp

```diff
@@ -0,0 +1,351 @@
+    bool          firstLabel = true;
+    uint8_t       labelLength;
+
+    while (true)
+    {
+        error = iterator.GetNextLabel();
+
+        switch (error)
+        {
+        case OT_ERROR_NONE:
+
+            if (!firstLabel)
+            {
+                *aNameBuffer++ = kLabelSeperatorChar;
+                aNameBufferSize--;
+
+                // No need to check if we have reached end of the name buffer
+                // here since `iterator.ReadLabel()` would verify it.
+            }
+
+            labelLength = static_cast<uint8_t>(OT_MIN(kMaxLabelLength + 1, aNameBufferSize));
+            SuccessOrExit(error = iterator.ReadLabel(aNameBuffer, labelLength, /* aAllowDotCharInLabel */ false));
+            aNameBuffer += labelLength;
+            aNameBufferSize -= labelLength;
+            firstLabel = false;
+            break;
+
+        case OT_ERROR_PENDING:
+            break;
+
+        case OT_ERROR_NOT_FOUND:
+            // We reach the end of name successfully. Always add a terminating dot
+            // at the end.
+            *aNameBuffer++ = kLabelSeperatorChar;
+            aNameBufferSize--;
+            VerifyOrExit(aNameBufferSize >= sizeof(uint8_t), error = OT_ERROR_NO_BUFS);
+            *aNameBuffer = kNullChar;
+            aOffset      = iterator.mNameEndOffset;
+            error        = OT_ERROR_NONE;
+
+            // Fall through
+
+        default:
+            ExitNow();
+        }
+    }
+
+exit:
+    return error;
+}
+
+otError DomainName::LabelIterator::GetNextLabel(void)
+{
+    otError  error;
+    uint8_t  labelLength;
+    uint16_t pointerValue;
+
+    SuccessOrExit(error = mMessage.Read(mNextLabelOffset, labelLength));
+
+    switch (labelLength & kLabelTypeMask)
+    {
+    case kTextLabelType:
+
+        if (labelLength == 0)
+        {
+            // Zero label length indicates end of a name.
+
+            if (!IsEndOffsetSet())
+            {
+                mNameEndOffset = mNextLabelOffset + sizeof(uint8_t);
+            }
+
+            ExitNow(error = OT_ERROR_NOT_FOUND);
+        }
+
+        mLabelStartOffset = mNextLabelOffset + sizeof(uint8_t);
+        mLabelLength      = labelLength;
+        mNextLabelOffset  = mLabelStartOffset + labelLength;
+        break;
+
+    case kPointerLabelType:
+
+        // A pointer label takes the form of a two byte sequence as a
+        // `uint16_t` value. The first two bits are ones. The next 14 bits
+        // specify an offset value from the start of the DNS header.
+
+        SuccessOrExit(error = mMessage.Read(mNextLabelOffset, pointerValue));
+
+        if (!IsEndOffsetSet())
+        {
+            mNameEndOffset = mNextLabelOffset + sizeof(uint16_t);
+        }
+
+        mNextLabelOffset = mHeaderOffset + (HostSwap16(pointerValue) & kPointerLabelOffsetMask);
+
+        // `OT_ERROR_PENDING` is used to indicate that we reached
+        // a pointer label.
+
+        error = OT_ERROR_PENDING;
```

Comment:
Can we follow pointer labels until we find a text label (in this method) so that we can avoid returning `OT_ERROR_PENDING`? 
I think this can simplify the callers of `GetNextLabel`.

Reply:
This is currently used in `ParseName()` where we want to get the offset to the end of the encoded name field. For this, we want to stop if we reach a pointer label (we don't care about any next label in this case).

We can continue reading the next label from `GetNextLabel()` and check in `ParseName()` if `mEndOffset` is set on `LabelIterator` (which indicate whethere we have reached the end of the name field and passed over a pointer label).
The downside I see is that we would do a bit extra work if we do this during `ParseName()` but I agree it may potentially simplify the code in other cases. Let me try it out and see how it looks. Good suggestion. Thanks. 👍 

For return type, I think we need to return `otError` since we can have `PARSE_ERROR` as well (e.g., invalid label type or message ends abruptly and does not contain the label field and/or pointer value at the given offset).

Reply:
> The downside I see is that we would do a bit extra work if we do this during ParseName()

The extra work would very likely to be just one more loop to find the real text label, which shouldn't be significant. 

I am fine to keep `otError` as the return value.

Reply:
Changed in new pushed commit.